engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java

```diff
@@ -1119,147 +1288,212 @@ public Class<?> visit(ArrayAccessExpr n, VisitArgs printer) {
+            final Node origParent = n.getParentNode().orElseThrow();
+            final MethodCallExpr newExpr = new MethodCallExpr(n.getName(), "get", NodeList.nodeList(n.getIndex()));
+            replaceChildExpression(origParent, n, newExpr);
+            nothingPrintedAssertion.run();
+            return newExpr.accept(this, printer);
         }
     }
 
     @Override
-    public Class<?> visit(BinaryExpr n, VisitArgs printer) {
+    public Class<?> visit(final BinaryExpr n, VisitArgs printer) {
+        final Runnable assertNothingPrinted = getNothingPrintedAssertion(printer);
+
         BinaryExpr.Operator op = n.getOperator();
+        final Node origParent = n.getParentNode().orElseThrow();
+
+        Class<?> lhType = getTypeWithCaching(n.getLeft());
+        Class<?> rhType = getTypeWithCaching(n.getRight());
+
         final Expression leftExpr = n.getLeft();
         final Expression rightExpr = n.getRight();
 
-        Class<?> lhType = getTypeWithCaching(leftExpr);
-        Class<?> rhType = getTypeWithCaching(rightExpr);
-
         if ((lhType == String.class || rhType == String.class) && op == BinaryExpr.Operator.PLUS) {
-            if (printer.hasStringBuilder()) {
-                leftExpr.accept(this, printer);
-                printer.append(getOperatorSymbol(op));
-                rightExpr.accept(this, printer);
-            }
+
+            leftExpr.accept(this, printer);
+            printer.append(' ').append(getOperatorSymbol(op)).append(' ');
+            rightExpr.accept(this, printer);
+
             return String.class;
         }
 
-        if (op == BinaryExpr.Operator.OR || op == BinaryExpr.Operator.AND) {
-            if (printer.hasStringBuilder()) {
-                leftExpr.accept(this, printer);
-                printer.append(getOperatorSymbol(op));
-                rightExpr.accept(this, printer);
-            }
+        // For boolean operations that don't involve Deephaven's ordering guarantees or null handling,
+        // the operation can be printed as-is instead of using QueryLanguageFunctionUtil methods.
+        final boolean isPrimitiveBooleanOperation =
+                boolean.class.equals(lhType) && boolean.class.equals(rhType) && (op == BinaryExpr.Operator.EQUALS
+                        || op == BinaryExpr.Operator.NOT_EQUALS
+                        || op == BinaryExpr.Operator.BINARY_OR
+                        || op == BinaryExpr.Operator.BINARY_AND
+                        || op == BinaryExpr.Operator.XOR);
+        if (op == BinaryExpr.Operator.OR || op == BinaryExpr.Operator.AND || isPrimitiveBooleanOperation) {
+
+            leftExpr.accept(this, printer);
+            printer.append(' ').append(getOperatorSymbol(op)).append(' ');
+            rightExpr.accept(this, printer);
+
             return boolean.class;
         }
 
+        // "Replace "x!=y" with "!(x==y)")
```

Comment:
Fixed the weird punctuation, and added an explanation. (We rewrite it so that it gets turned into `!eq(x, y)` — there is no `neq()`in `QueryLanguageFunctionUtils`.)